lib/backgroundscanner.php

```diff
@@ -110,22 +131,43 @@ public function run(){
 		\OC_Util::tearDownFS();
 	}
 
+	protected function getOwner($fileId){
+		$mountProviderCollection = \OC::$server->getMountProviderCollection();
+		$mountCache = $mountProviderCollection->getMountCache();
+		$mounts = $mountCache->getMountsForFileId($fileId);
+		if (!empty($mounts)) {
+			$user = $mounts[0]->getUser();
+			if ($user instanceof IUser) {
+				return $user;
+			}
+		}
+	}
+
 	/**
-	 * A hack to access files and views. Better than before.
-	 *
-	 * @return bool
+	 * @param \OCP\IUser $user
+	 * @return \OCP\Files\Folder
 	 */
-	protected function initFS(){
-		//Need any valid user to mount FS
-		$results = $this->userManager->search('', 2, 0);
-		$anyUser = array_pop($results);
-		if (is_null($anyUser)) {
-			\OC::$server->getLogger()->error("Failed to setup file system", ['app' => 'files_antivirus']);
-			return false;
+	protected function getUserFolder(IUser $user) {
+		if (!isset($this->userFolders[$user->getUID()])) {
+			$userFolder = $this->rootFolder->getUserFolder($user->getUID());
+			$this->userFolders[$user->getUID()] = $userFolder;
+		}
+		return $this->userFolders[$user->getUID()];
+	}
+
+	/**
+	 * @param IUser $user
+	 */
+	protected function initFilesystemForUser(IUser $user) {
+		if ($this->currentFilesystemUser !== $user->getUID()) {
+			if ($this->currentFilesystemUser !== '') {
+				Filesystem::tearDown();
+			}
+			Filesystem::init($user->getUID(), '/' . $user->getUID() . '/files');
+			$this->userSession->setUser($user);
```

Comment:
at you setting this back to null at some point ? Because multiple bkg jobs could run in a single pass so we don't want a stray user when running the next job...